src/fsharp/FSharp.Compiler.Private/Utilities/ShardMap.fs

```diff
@@ -0,0 +1,1504 @@
+            if map.BucketSize > bucket.Length then
+                let target = Array.zeroCreate<Shard<'K,'V>>(map.BucketSize)
+                Array.Copy(map.getBucket(),target,map.BucketSize)
+                target, bucket, ref (map.Count)
+            else
+                let target = Array.zeroCreate<Shard<'K,'V>>(bucket.Length)
+                Array.Copy(bucket,target,bucket.Length)
+                target, map.getBucket() , ref (!countRef)
+                
+        let trgtBitDepth = calcBitMaskDepth ((target.Length * ShardSize) - 1)
+        let bucketBitMask = calcSubBitMask trgtBitDepth
+        
+        let mapFn = 
+            if additions.Length = target.Length then
+                // Split into two types of mapping, one-to-one, and small map expansion
+                ////////////////////////////////////////////////////////////////////////
+                fun bi si sm ->
+                    let tshrd = target.[bi]
+                    let tm = tshrd.[si]
+                    // Buckets same size so simple to one-to-one map
+                    if isNull tm then
+                        Interlocked.Add(tCount,MapTree.size sm) |> ignore
+                        tshrd.[si] <- sm
+                    else
+                        tshrd.[si] <-
+                            MapTree.fold (fun acc k v -> // for each key in source
+                                MapTree.addAndIncr comparer k v tCount acc // lookup doubled so might be worth extending Maptree to return both tree & `newAddition' bool
+                            ) tm sm
+            else
+                fun _ si sm ->
+                    // Bucket lower order so shard needs to be remapped to higher order
+                    ///////////////////////////////////////////////////////////////////////
+                    MapTree.iter (fun k v -> // for each key in source
+                        let kh = k.GetHashCode()
+                        let tbi = bucketIndex(kh,bucketBitMask)
+                        let tshrd = target.[tbi]
+                        let tm = tshrd.[si]
+
+                        if isNull tm then
+                            Interlocked.Increment(tCount) |> ignore
+                            tshrd.[si] <- MapOne(k,v)
+                        else                                             
+                            tshrd.[si] <- MapTree.addAndIncr comparer k v tCount tm
+                    ) sm
+
+
+        eachMapInBucket(additions , mapFn)
+        // Step through maps to union
+        ShardMap<'K,'V>(!tCount,target)
+
+
+    /////////////////////////////////
+    /// Static Methods
+    /////////////////////////////////
+
+    static member inline Empty with get () = ShardMap<'K,'V>(0)
+
+    static member Collect (collectFn: 'T -> #seq<'V>) (keyFn:'V -> 'K) (collection:seq<'T>) =
+        
+        let initSize = Seq.length collection //:HACK
```

Comment:
this iterates the collection?

Reply:
This was removed (and hack as per comment) but in PR due to merging issue apologies, please disregard